accepted/2021/System.Diagnostics/Metrics-Design.md

```diff
@@ -0,0 +1,511 @@
+                            T measurement,
+                            ReadOnlySpan<KeyValuePair<string, object?>> tags,
+                            object? cookie);
+
+
+    /// <summary>
+    /// The listener class can be used to listen to observable and non-observable instrument
+    /// recorded measurements.
+    /// </summary>
+    public class MeterListener : IDisposable
+    {
+        /// <summary>
+        /// Simple constructor
+        /// </summary>
+        public MeterListener() { throw null;  }
+
+        /// <summary>
+        /// Callbacks to get notification when an instrument is published
+        /// </summary>
+        public Action<Instrument>? InstrumentPublished { get; set; }
+
+        /// <summary>
+        /// Callbacks to get notification when stopping the measurement on some instrument
+        /// this can happen when the Meter or the Listener is disposed of. Or calling Stop()
+        /// on the listener.
+        /// </summary>
+        // This need some clarification
+        public Action<Instrument, object?>? MeasurementsCompleted { get; set; }
+
+        /// <summary>
+        /// Start listening to a specific instrument measurement recording.
+        /// </summary>
+        public void EnableMeasurementEvents(Instrument instrument, object? cookie = null) { throw null; }
+
+        /// <summary>
+        /// Stop listening to a specific instrument measurement recording.
+        /// </summary>
+        public void DisableMeasurementEvents(Instrument instrument) { throw null; }
+
+        /// <summary>
+        /// Set a callback for a specific numeric type to get the measurement recording notification
+        /// from all instruments which enabled listened to and was created with the same specified
+        /// numeric type.
+        /// </summary>
+        public void SetMeasurementEventCallback<T>(MeasurementCallback<T>? measurementCallback) { throw null; }
+
+        public void Start() { throw null; }
+        public void Stop() { throw null; }
+
+        /// <summary>
+        /// Call all Observable instruments to get the recorded measurements reported to the
+        /// callbacks enabled by SetMeasurementEventCallback<T>
+        /// </summary>
+        public void RecordObservableInstruments() { throw null; }
+
+        public void Dispose() { throw null; }
+    }
+}
+```
+
+### Library Measurement Recording Example
+
+```csharp
+    Meter meter = new Meter("io.opentelemetry.contrib.mongodb", "v1.0");
+    Counter<int> counter = meter.CreateCounter<int>("Requests");
+    counter.Add(1);
+    counter.Add(1, KeyValuePair<string, object>("request", "read"));
+```
+
+### Listening Example
+
+```csharp
+    InstrumentListener listener = new InstrumentListener();
+    listener.InstrumentPublished = (instrument) =>
+    {
+        if (instrument.Name == "Requests" && instrument.Meter.Name == "io.opentelemetry.contrib.mongodb")
+        {
+            listener.EnableMeasurementEvents(instrument, null);
```

Comment:
nit: Ideally, we don't have to capture listener in this closure.

Reply:
Actually, @noahfalk had it before passing the MeterListener object to the `InstrumentPublished` callback and we thought it is not really needed. Thinking more about and expecting users will need to call `EnableMeasurementEvents` inside the `InstrumentPublished`, I have returned it back so users can avoid capture the listener in the closure.